File: python/src/solana_keychain/gcp_kms/signer.py
"""Google Cloud KMS signer integration using EdDSA (Ed25519) signing."""

from dataclasses import dataclass, field
from typing import Any

try:
    from google.api_core.exceptions import GoogleAPIError
    from google.auth.exceptions import GoogleAuthError
    from google.cloud import kms_v1
except ImportError as error:  # pragma: no cover
    raise ImportError(
        "solana_keychain.gcp_kms requires the gcp-kms extra: pip install 'solana-keychain[gcp-kms]'"
    ) from error

from solders.pubkey import Pubkey
from solders.signature import Signature
from solders.transaction import Transaction

from solana_keychain.core.errors import SignerError, SignerErrorCode
from solana_keychain.core.signer import SignedTransaction, SolanaSigner
from solana_keychain.core.transaction_util import (
    ED25519_SIGNATURE_LENGTH,
    add_signature_to_transaction,
    classify_signed_transaction,
    serialize_transaction,
)

EC_SIGN_ED25519 = kms_v1.CryptoKeyVersion.CryptoKeyVersionAlgorithm.EC_SIGN_ED25519


@dataclass
class GcpKmsSignerConfig:
    """Configuration for a GCP KMS signer.

    ``key_name`` is the full resource name of the crypto key version, e.g.
    ``projects/p/locations/l/keyRings/r/cryptoKeys/k/cryptoKeyVersions/1``, and must
    reference an ``EC_SIGN_ED25519`` key. ``public_key`` is the base58 Solana public
    key corresponding to that key. ``client`` accepts a pre-configured
    ``KeyManagementServiceAsyncClient`` (custom credentials, endpoint, or transport).
    """

    key_name: str
    public_key: str
    client: Any | None = field(default=None, repr=False)


class GcpKmsSigner(SolanaSigner):
    """Signer backed by a Google Cloud KMS Ed25519 key."""

    def __init__(self, config: GcpKmsSignerConfig) -> None:
        try:
            self._pubkey = Pubkey.from_string(config.public_key)
        except Exception:
            raise SignerError(SignerErrorCode.INVALID_PUBLIC_KEY, "Invalid public key") from None
        self._key_name = config.key_name
        self._client = (
            config.client if config.client is not None else kms_v1.KeyManagementServiceAsyncClient()
        )

    def __repr__(self) -> str:
        return f"GcpKmsSigner(key_name={self._key_name}, pubkey={self._pubkey})"

    @property
    def key_name(self) -> str:
        return self._key_name

    @property
    def pubkey(self) -> Pubkey:
        return self._pubkey

    async def _sign_bytes(self, message: bytes) -> Signature:
        # EC_SIGN_ED25519 operates in PureEdDSA mode: the request carries the raw
        # data, never a digest.
        try:
            response = await self._client.asymmetric_sign(
                request={"name": self._key_name, "data": message}
            )
        except (GoogleAPIError, GoogleAuthError) as error:
            raise SignerError(
                SignerErrorCode.REMOTE_API_ERROR, f"GCP KMS Sign operation failed: {error}"
            ) from None
        signature_bytes = bytes(response.signature)
        if not signature_bytes:
            raise SignerError(SignerErrorCode.SIGNING_FAILED, "No signature in GCP KMS response")
        if len(signature_bytes) != ED25519_SIGNATURE_LENGTH:
            raise SignerError(
                SignerErrorCode.SIGNING_FAILED,
                f"Invalid signature length: expected {ED25519_SIGNATURE_LENGTH} bytes, "
                f"got {len(signature_bytes)}",
            )
        signature = Signature.from_bytes(signature_bytes)
        if not signature.verify(self._pubkey, message):
            raise SignerError(
                SignerErrorCode.SIGNING_FAILED,
                "Signature verification failed — the returned signature does not match "
                "the public key",
            )
        return signature

    async def sign_transaction(self, transaction: Transaction) -> SignedTransaction:
        signature = await self._sign_bytes(transaction.message_data())
        add_signature_to_transaction(transaction, self._pubkey, signature)
        return classify_signed_transaction(
            transaction, serialize_transaction(transaction), signature
        )

    async def sign_message(self, message: bytes) -> Signature:
        return await self._sign_bytes(message)

    async def is_available(self) -> bool:
        try:
            response = await self._client.get_public_key(request={"name": self._key_name})
        except (GoogleAPIError, GoogleAuthError):
            return False
        return bool(response.algorithm == EC_SIGN_ED25519)


async def create_gcp_kms_signer(config: GcpKmsSignerConfig) -> GcpKmsSigner:
    """Create a ready-to-use GCP KMS signer.

    The client is constructed on the event loop: the gRPC async transport binds to
    the running loop, so construction must not be moved to a worker thread.
    """
    return GcpKmsSigner(config)

File: go/signers/gcpkms/signer.go
package gcpkms

import (
	"context"

	kms "cloud.google.com/go/kms/apiv1"
	"cloud.google.com/go/kms/apiv1/kmspb"
	"github.com/gagliardetto/solana-go"

	"github.com/solana-foundation/solana-keychain/go/core/v2"
)

// Signer signs with an Ed25519 key held in Google Cloud KMS. AsymmetricSign is
// invoked in PureEdDSA mode with the raw message bytes, and every returned
// signature is verified against the configured public key before being
// surfaced.
//
// A Signer is immutable after New and safe for concurrent use.
type Signer struct {
	client  KMSClient
	keyName string
	pubkey  solana.PublicKey
}

// Ensure Signer satisfies the core contract at compile time.
var _ core.Signer = (*Signer)(nil)

// String renders the signer without config detail, so fmt's reflective struct
// printing cannot surface the key resource name or client internals.
func (s Signer) String() string {
	return "gcpkms.Signer{pubkey: " + s.pubkey.String() + "}"
}

// GoString mirrors String so %#v cannot leak either.
func (s Signer) GoString() string { return s.String() }

// New builds a GCP KMS signer from cfg. When cfg.Client is nil it dials the
// official KMS client (I/O), so the returned signer is ready to use.
func New(ctx context.Context, cfg Config) (*Signer, error) {
	pubkey, err := solana.PublicKeyFromBase58(cfg.PublicKey)
	if err != nil {
		return nil, core.WrapSignerError(core.CodeInvalidPublicKey, "invalid public key", err)
	}

	client := cfg.Client
	if client == nil {
		c, err := kms.NewKeyManagementClient(ctx, cfg.ClientOptions...)
		if err != nil {
			return nil, core.WrapSignerError(core.CodeRemoteAPIError, "failed to create KMS client", err)
		}
		client = c
	}

	return &Signer{client: client, keyName: cfg.KeyName, pubkey: pubkey}, nil
}

// Pubkey returns the signer's Solana public key.
func (s *Signer) Pubkey() solana.PublicKey { return s.pubkey }

// KeyName returns the full resource name of the crypto key version.
func (s *Signer) KeyName() string { return s.keyName }

// Close releases the underlying KMS client's connections, including a client
// supplied via Config.Client — callers sharing one client across signers should
// close it themselves instead of calling Close here.
func (s *Signer) Close() error { return s.client.Close() }

// signBytes signs message with GCP KMS EdDSA signing. EC_SIGN_ED25519 operates
// in PureEdDSA mode, so the raw message bytes go in the request's data field
// (not a digest).
func (s *Signer) signBytes(ctx context.Context, message []byte) (solana.Signature, error) {
	resp, err := s.client.AsymmetricSign(ctx, &kmspb.AsymmetricSignRequest{
		Name: s.keyName,
		Data: message,
	})
	if err != nil {
		return solana.Signature{}, core.WrapSignerError(core.CodeRemoteAPIError, "GCP KMS Sign operation failed", err)
	}

	sigBytes := resp.GetSignature()
	if len(sigBytes) == 0 {
		return solana.Signature{}, core.NewSignerError(core.CodeSigningFailed, "no signature in GCP KMS response")
	}
	sig, err := core.SignatureFromBytes(sigBytes, "gcp kms")
	if err != nil {
		return solana.Signature{}, err
	}

	if err := core.VerifySignature(s.pubkey, message, sig); err != nil {
		return solana.Signature{}, err
	}
	return sig, nil
}

// SignMessage signs arbitrary bytes with the KMS-held key.
func (s *Signer) SignMessage(ctx context.Context, message []byte) (solana.Signature, error) {
	return s.signBytes(ctx, message)
}

// SignTransaction signs the transaction's message bytes and inserts the
// signature at this signer's required-signer position, returning the encoded
// transaction and its completeness.
func (s *Signer) SignTransaction(ctx context.Context, tx *solana.Transaction) (core.SignedTransaction, error) {
	return core.SignTransactionWith(ctx, tx, s.pubkey, s.signBytes)
}

// IsAvailable reports whether the crypto key version is reachable and uses the
// EC_SIGN_ED25519 algorithm. All errors are swallowed and reported as false.
func (s *Signer) IsAvailable(ctx context.Context) bool {
	resp, err := s.client.GetPublicKey(ctx, &kmspb.GetPublicKeyRequest{Name: s.keyName})
	if err != nil {
		return false
	}
	return resp.GetAlgorithm() == kmspb.CryptoKeyVersion_EC_SIGN_ED25519
}
